File: 0005-longest-palindromic-substring/solution.java
class Solution {
    public String longestPalindrome(String s) {
        int max = -1;
        String longest = null;
        boolean[][] dp = new boolean[s.length()][s.length()];
        for(int i=s.length()-1; i>=0; i--) {
            for(int j=i; j<s.length(); j++) {
                dp[i][j] = s.charAt(i)==s.charAt(j)&&((j-i+1<3)||dp[i+1][j-1]);
                if((dp[i][j])&&(j-i+1>=max)) {
                    max=j-i+1;
                    longest=s.substring(i,j+1);
                }
            }
        }
        return longest;
    }
}

File: 0005-longest-palindromic-substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:
            return s
        dp = [
            [
                1 if i >= j else None
                for j in range(len(s))
            ]
            for i in range(len(s))
        ]
        max_len = 1
        ret = s[0]
        for offset in range(1, len(s)):
            for i in range(0, len(s)-offset):
                j = i + offset
                if s[i] == s[j]:
                    dp[i][j] = dp[i+1][j-1]
                else:
                    dp[i][j] = 0
                if dp[i][j] == 1 and j-i+1 > max_len:
                    max_len = j-i+1
                    ret = s[i:j+1]
        return ret


